File: 41.缺失的第一个正数.go
/*
 * @lc app=leetcode.cn id=41 lang=golang
 *
 * [41] 缺失的第一个正数
 */
package main

// @lc code=start
func firstMissingPositive(nums []int) int {
	n := len(nums)
	for i := 0; i < n; i++ {
		if nums[i] <= 0 {
			nums[i] = n + 1
		}
	}
	for i := 0; i < n; i++ {
		num := myNewAbs(nums[i])
		if num <= n {
			nums[num-1] = -myNewAbs(nums[num-1])
		}
	}
	for i := 0; i < n; i++ {
		if nums[i] > 0 {
			return i + 1
		}
	}
	return n + 1
}

func myNewAbs(x int) int {
	if x < 0 {
		return -x
	}
	return x
}

// @lc code=end

File: 41.缺失的第一个正数.rs
/*
 * @lc app=leetcode.cn id=41 lang=rust
 *
 * [41] 缺失的第一个正数
 */

// @lc code=start
impl Solution {
    pub fn first_missing_positive(nums: Vec<i32>) -> i32 {
        let mut nums = nums;
        let len = nums.len();
        if len == 0 {
            return 0;
        }
        for i in 0..len {
            if nums[i] <= 0 {
                nums[i] = len as i32 + 1;
            }
        }
        for i in 0..len {
            let num = i32::abs(nums[i]);
            if num as usize <= len {
                nums[num as usize - 1] = -i32::abs(nums[num as usize - 1])
            }
        }
        for i in 0..len {
            if nums[i] >= 0 {
                return i as i32 + 1;
            }
        }
        len as i32 + 1
    }
}
// @lc code=end

